typewriter/fixes/fixer_utils.py

```diff
@@ -133,3 +140,254 @@ def touch_import(package, name, node):
+            if _find(name, child.children[0]):
+                ret = child
+
+        if ret:
+            if not package:
+                return ret
+            if is_import(ret):
+                return ret
+    return None
+
+
+def find_import_info(package, name, node):
+    # type: (Optional[str], str, Node) -> Optional[ImportInfo]
+    """
+    Finds the import statement for <name> regardless of whether <name> is the binding name
+        ex where name='a':
+        import a => match AND import a as b => match BUT import b as a => None
+
+    Parameters
+    -----------
+    package : Optional[str]
+    name : str
+    node : Node
+
+    Return
+    -----------
+    Optional[ImportInfo]
+    """
+    for child in node.children:
+        if is_import(child):
+            ret = get_import_info(child)
+            if ret.package == package and dict(ret.imports).get(name) is not None:
+                return ret
+        elif child.type == syms.simple_stmt:
+            return find_import_info(package, name, child)
+    return None
+
+
+def decompose_name(node):
+    """
+    NOTE: Per the lib2to3 grammar:
+            dotted_name: NAME ('.' NAME)*
+          This means that dotted_name can be either dotted or not dotted, i.e. it's a generalized
+          form of NAME. So this function will cover both cases.
+
+    Given a dotted_name node this will return a tuple of the form (pkg, name, full_string) where
+    all are str
+        ex: a.b.c => (a.b, c, a.b.c)
+            b.c => (b, c, b.c)
+            c => (None, c, c)
+    otherwise it will return None for each field
+    """
+    if node.type == token.NAME:
+        # node is just a name, no dots
+        return None, node.value, node.value
+
+    if node.children:
+        # Right most node will be the name, i.e. a.b.c = ['a','.','b','.','c']
+        name_node = node.children[-1]
+        package_nodes = node.children[:-2]
+        name = str(name_node).strip()
+        package = ''.join(str(n).strip() for n in package_nodes)
+        full = ''.join(str(n).strip() for n in node.children)
+        return package, name, full
+
+    return None, None, None
+
+
+def get_import_info(node):
+    """
+    Wraps the cachable get_import_info in order to convert the node into a hashable node before
+    attempting to cache
+    """
+    node = HashableNode(node)
+    return _get_import_info_cachable(node)
+
+
+class HashableNode(object):
+    """
+    Hashable wrapper for node and leaf objects in order to cache
+    """
+
+    def __init__(self, node):
+        # type: (Union[Node, Leaf]) -> None
+        """
+        Parameters
+        -----------
+        node : Union[Node, Leaf]
+        """
+        self.node = node
+
+    def __hash__(self):
+        return hash((self.node.get_lineno(),
+                     self.node.depth(),
+                     self.node.__str__()))
+
+
+@cache
+def _get_import_info_cachable(hashable_node):
+    # type: (HashableNode) -> ImportInfo
```

Comment:
I think the return type here needs to be `List[ImportInfo]` - since a single import statement can yield results from multiple packages - ie, `import a.b.c, d.e.f`.

Reply:
I opted to make it a dictionary keyed by package name (with no package being labeled 'no_pkg'). For searching over results this is more efficient and intuitive (I think) than making it a list. Let me know if you disagree